### PR TITLE
Make the RingSolver work with generic rings (#1116)

### PR DESCRIPTION
**Pull request by felixwellen**
Work on issue #1116 . So far, this only concerns the problem that deBruijn-indices have to be adjusted when using the forall-version of the solver macro.

### CODE REVIEW

File: src/Reflection/AST/DeBruijn.agda [hunk 1]
```diff
@@ -53,6 +55,16 @@ module _ where
   weakenClauses : (by : ℕ) → Clauses → Clauses
   weakenClauses = weakenFrom′ traverseClauses 0
 
+-- Apply Weakening to substitute under lambdas
+
+prependLams : List (String × Visibility) → Term → Term
+prependLams xs t = foldr (λ {(s , v) t → lam v (abs s t)}) t (reverse xs)
```
Comment (gallais) [inline src/Reflection/AST/DeBruijn.agda:61]: Note that you don't need a pattern-matching lambda for irrefutable patterns like `(s , v)`

File: src/Tactic/RingSolver/Examples.agda [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  ℤAsCRing = record
+               { Carrier = ℤ
+               ; _≈_ = _
+               ; _+_ = _
+               ; _*_ = _
+               ; -_ = _
+               ; 0# = _
+               ; 1# = _
+               ; isCommutativeRing = +-*-isCommutativeRing
+               }
```
Comment (gallais) [inline src/Tactic/RingSolver/Examples.agda:34]:
```suggestion
               { Carrier = ℤ
               ; isCommutativeRing = +-*-isCommutativeRing
               }
```

File: src/Reflection/AST/DeBruijn.agda (hunk 1 above)
Comment (felixwellen) [reply src/Reflection/AST/DeBruijn.agda:61]: Thanks! Didn't know that..

File: src/Tactic/RingSolver/Examples.agda [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+open import Data.Integer.Properties
+open import Data.Maybe
+
+open import Agda.Builtin.Equality using (_≡_ ; refl)
+
+open import Tactic.RingSolver
+open import Tactic.RingSolver.Core.AlmostCommutativeRing
+
+
+module IntegerExamples where
```
Comment (MatthewDaggitt) [inline src/Tactic/RingSolver/Examples.agda:23]: You could add these examples to the existing [README.Tactic.RingSolver](https://github.com/agda/agda-stdlib/blob/master/README/Tactic/RingSolver.agda)?

**Comment by MatthewDaggitt** [comment]
@felixwellen unless you have any objections in the next week or so I'll close this PR as it's going stale?

**Comment by felixwellen** [comment]
Yes, I think it can be closed. From my vague memories, the status was, that I didn't understand why some expressions do not normalize and I don't think it is worth figuring out.

**Pull request closed without merging**